### ISSUE TITLE
Event directives on built in components

### ISSUE THREAD

**Issue by zamith** [msaraiva/surface#82]
I would assume that this would work:

```elixir  
alias Surface.Components.Link
<Link to="#" :on-phx-click={{ "ok", target: "#comp" }}>OK</Link>
```

But it fails to compile and emits a warning.

```
Unknown property ":on-phx-click" for component <Link>
```

How can I make it work?

**Comment by msaraiva** [comment]
Hi @zamith!

I just added the `click` event to `<Link>` in de6d22ffe5ddf4cd149069c82d1fa33adb688ec4. 

Ideally, we should add all other valid events. Until then, in case you need any other event, just use the `event_to_opts/2` function to properly convert Surface's events into a keyword list that can be appended to the helper's `opts`.

@alexandrubagu we need to forward the events in the **form helpers** using `event_to_opts/2` too because they will not be passed automatically. We can later see a way to group the events definitions in some macros so we can add them easily.

**Comment by zamith** [comment]
Any reason why the event directives can only be applied to tags and not components, unlike the other directives?

**Comment by zamith** [comment]
Also, shall we do the same for `phx-submit` for Form? I'm trying to use `event_to_opts` and it doesn't seem to work for this scenario:

```elixir
<Form for={{ :query }} action="#" opts={{ event_to_opts(@submit, :phx_submit) }}>
```

I get a `no function clause matching in Phoenix.HTML.Tag.build_attrs/3`, because the `opts` are passed in as `[[phx_submit: "search"]]`. I'm assuming this has to do with that syntactic sugar around keywords.

**Comment by msaraiva** [comment]
Hi @zamith!

> Any reason why the event directives can only be applied to tags and not components, unlike the other directives?

Because when using `:on-phx-click` on a HTML tag, we know for sure that we should add `phx-click` and `phx-target` directly to its attributes. When using a component, there's no way we know what should be done. For instance, imagine you have a component composed with a parent `div`, some other markup and a `button`:

```elixir
  def render(assigns) do
    ~H"""
    <div>
       ...
      <button ...>Ok</button>
    </div>
    """
  end
```

if I could add a `:on-phx-click` when using this component, which tag should be updated? The div or the button? My intention could be the button, but there's no way Surface can infer that.

Another example is when wrapping helpers as we did for the inputs. We don't render any markup directly. We just call a function an pass arguments to it. We could have called more than one function or added other components/markup. Again, there's no way Surface knows what should be done.

> I get a no function clause matching in Phoenix.HTML.Tag.build_attrs/3, because the opts are passed in as [[phx_submit: "search"]]. I'm assuming this has to do with that syntactic sugar around keywords.

Yup. Good catch! I'll fix that.

**Comment by msaraiva** [comment]
@zamith regarding the `<Form>` events, I think we should follow the same approach we did for `<Link>` which would be to explicitly add the events to the Form's public API:

```elixir
  @doc "Triggered on submit"
  property submit, :event  

  def init_context(assigns) do
    submit_opts = event_to_opts(assigns.submit, :phx_submit)
    form = form_for(assigns.for, assigns.action, assigns.opts ++ submit_opts)
    {:ok, form: form}
  end
```

and then use it as:

```jsx
<Form for={{ :query }} action="#" submit={{@submit}}>
```

We need to this for all valid events. As you can see [here](https://github.com/msaraiva/surface/pull/79/files#diff-fe6b27ec5ce46e7faa9480653166f015R17-R22), @alexandrubagu already declared the events. The only thing missing the passing them to the underlying helper.

**Comment by msaraiva** [comment]
@zamith the issue related to `no function clause matching in Phoenix.HTML.Tag.build_attrs/3` should have been fixed in a31988842acad51ced1b30ada4c963b67f06446a. Please, let me know if you still find any issue.

Thanks for reporting the problem 👍
